Code/PgSQL: Fix Pointer vs Datum (Compatibility with PG16)

df7cb opened the pull request:
PostgreSQL 16 got stricter wrt proper Pointer and Datum handling.

Fix developed against 202303.3 and tested with PostgreSQL 10 up to 16.

greglandrum commented:
@rvianello do these changes look ok to you?

df7cb commented:
Hi, it was a hard failure here, perhaps in conjunction with stricter compiler flags on Debian unstable.

Will check again tomorrow, too late now to dig up build logs.

rvianello commented:
ah, maybe I got it now.. PG_DETOAST_DATUM returns a struct varlena *, it was ok as long as DatumGetPointer was a macro just casting to a different pointer type, but now it's a function expecting a Datum (a long unsigned) as argument, and this results in making an "integer from pointer without a cast". Sorry, I missed this warning yesterday.

but don't we need the same change at bfp_gist.c:657,659 and rdkit_gist:549,550 then?

df7cb commented:
Here's the error message without the patch, PG 16.0 on Debian unstable/amd64. clang means this step happens when compiling the .bc files for JIT support:

```
make[1]: Entering directory '/home/myon/projects/postgresql/rdkit/rdkit/build-16'
/usr/bin/clang-15 -Wno-ignored-attributes -fno-strict-aliasing -fwrapv -Xclang -no-opaque-pointers -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -O2  -I/usr/local/include -I/home/myon/projects/postgresql/rdkit/rdkit/Code -DRDKITVER='"007300"' -DRDK_TOOLKIT_VERSION='"2022.09.01"'   -I. -I/home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit -I/usr/include/postgresql/16/server -I/usr/include/postgresql/internal  -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -flto=thin -emit-llvm -c -o rdkit_gist.bc /home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit/rdkit_gist.c
/home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit/rdkit_gist.c:121:41: error: incompatible pointer to integer conversion passing 'struct varlena *' to parameter of type 'Datum' (aka 'unsigned long') [-Wint-conversion]
  bytea *key = (bytea *)DatumGetPointer(PG_DETOAST_DATUM(entry->key));
                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/postgresql/16/server/fmgr.h:241:2: note: expanded from macro 'PG_DETOAST_DATUM'
        pg_detoast_datum((struct varlena *) DatumGetPointer(datum))
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/postgresql/16/server/postgres.h:312:23: note: passing argument to parameter 'X' here
DatumGetPointer(Datum X)
                      ^
1 error generated.
```

I've been developing this on the 202303.3 release tarball, it's possible that git HEAD needs more fixes.

df7cb commented:
... right, gcc is also raising warnings (but not errors):

```
/home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit/mol_op.c: In function ‘mol_murckoscaffold’:
/home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit/mol_op.c:343:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
  343 |   CROMol scaffold = MolMurckoScaffold(mol);
      |   ^~~~~~
...

gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wshadow=compatible-local -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -g -g -O2 -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -fno-omit-frame-pointer -g -O2 -ffile-prefix-map=/home/myon/projects/postgresql/rdkit/rdkit=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -fvisibility=hidden -I/usr/local/include -I/home/myon/projects/postgresql/rdkit/rdkit/Code -DRDKITVER='"007300"' -DRDK_TOOLKIT_VERSION='"2022.09.01"'   -I. -I/home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit -I/usr/include/postgresql/16/server -I/usr/include/postgresql/internal  -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -fPIC -c -o rdkit_gist.o /home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit/rdkit_gist.c
In file included from /home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit/rdkit_gist.c:33:
/home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit/rdkit_gist.c: In function ‘gmol_decompress’:
/usr/include/postgresql/16/server/fmgr.h:241:9: warning: passing argument 1 of ‘DatumGetPointer’ makes integer from pointer without a cast [-Wint-conversion]
  241 |         pg_detoast_datum((struct varlena *) DatumGetPointer(datum))
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         |
      |         struct varlena *
...

/usr/bin/clang-15 -Wno-ignored-attributes -fno-strict-aliasing -fwrapv -Xclang -no-opaque-pointers -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -O2  -I/usr/local/include -I/home/myon/projects/postgresql/rdkit/rdkit/Code -DRDKITVER='"007300"' -DRDK_TOOLKIT_VERSION='"2022.09.01"'   -I. -I/home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit -I/usr/include/postgresql/16/server -I/usr/include/postgresql/internal  -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -flto=thin -emit-llvm -c -o rxn_op.bc /home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit/rxn_op.c
/home/myon/projects/postgresql/rdkit/rdkit/Code/PgSQL/rdkit/rxn_op.c:100:37: warning: passing arguments to 'ReactionSubstructFP' without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
  PG_RETURN_BOOL(ReactionSubstructFP(rxn, rxn2));
                                    ^
...
```

rvianello commented:
I created a small PR against your branch to apply the same fix in a couple additional places. If those changes look ok to you, I think it would make sense to include them in this PR.

greglandrum commented:
@df7cb please @ mention me if you come back to this so that I see the update

df7cb commented:
@rvianello sorry I can't see that "small PR", where is it?

I rebased my pg16 branch, it still applied cleanly, and added the changes to bfp_gist.c:657,659 and rdkit_gist:549,550.

df7cb commented:
@greglandrum ^

rvianello commented:
@df7cb it was this one here https://github.com/df7cb/rdkit/pull/1 (you may have now implemented the same changes already, I didn't check).

rvianello commented:
@df7cb @greglandrum sorry, the conversation around this PR was a bit intermittent and I noticed the related warnings again today. I think the PR was ready to merge, or were there any pending questions?

greglandrum reviewed (approved):
LGTM based on Riccardo's comments

greglandrum commented:
Thanks @df7cb !
